File: string/2047_number_of_valid_words_in_a_sentence/2047_number_of_valid_words_in_a_sentence.py
class Solution:
    def countValidWords(self, sentence: str) -> int:
        result = 0
        for w in sentence.split():
            tmp = False
            for idx,c in enumerate(w):
                if str(c).isdigit():
                    break
                if c in '!.,' and idx != len(w)-1:
                        break
                if c == '-':
                    if tmp:
                        break
                    if idx == 0 or idx == len(w)-1:
                        break
                    if not w[idx-1].islower() or not w[idx+1].islower():
                        break
                    tmp = True
            else:
                result += 1
                tmp = False
        return result

File: string/2047_number_of_valid_words_in_a_sentence/2047_number_of_valid_words_in_a_sentence.go
package main

func valid(s string) bool {
	hasHyphens := false
	for i, ch := range s {
		if unicode.IsDigit(ch) || strings.ContainsRune("!.,", ch) && i < len(s)-1 {
			return false
		}
		if ch == '-' {
			if hasHyphens || i == 0 || i == len(s)-1 || !unicode.IsLower(rune(s[i-1])) || !unicode.IsLower(rune(s[i+1])) {
				return false
			}
			hasHyphens = true
		}
	}
	return true
}

func countValidWords(sentence string) (ans int) {
	for _, s := range strings.Fields(sentence) {
		if valid(s) {
			ans++
		}
	}
	return
}
